pandas/_libs/tslibs/conversion.pyx

```diff
@@ -347,9 +347,13 @@ cdef _TSObject convert_datetime_to_tsobject(datetime ts, object tz,
     if tz is not None:
         tz = maybe_get_tz(tz)
 
-        # sort of a temporary hack
         if ts.tzinfo is not None:
             if hasattr(tz, 'normalize') and hasattr(ts.tzinfo, '_utcoffset'):
+                # tz.localize does not correctly localize Timestamps near DST
+                # but correctly localizes datetimes
```

Comment:
IIRC `tz.localize` (or possibly `tz.normalize` if thats relevant) operates by calling `ts` methods.  Is this potentially something we can fix "higher up" in `Timestamp`?